crates/bls-crypto/src/bls/public.rs

```diff
@@ -216,25 +169,131 @@ impl CanonicalDeserialize for PublicKey {
 mod test {
     use super::*;
     use crate::bls::PrivateKey;
+    use algebra::{
+        bls12_377::{g2::Parameters as Bls12_377G2Parameters, Fq, Fq2},
+        curves::SWModelParameters,
+        fields::{Field, PrimeField, SquareRootField},
+    };
     use rand::thread_rng;
+    use std::io::{self, Result as IoResult, Write};
 
     #[test]
     fn test_public_key_serialization() {
         PublicKeyCache::resize(256);
         PublicKeyCache::clear_cache();
 
+        let old_serialization_logic = |pk: PublicKey, writer: &mut Vec<u8>| -> IoResult<_> {
+            let affine = pk.0.into_affine();
+            let mut x_bytes: Vec<u8> = vec![];
+            let y_c0_big = affine.y.c0.into_repr();
+            let y_c1_big = affine.y.c1.into_repr();
+            let half = Fq::modulus_minus_one_div_two();
+            affine.x.write(&mut x_bytes)?;
+            let num_x_bytes = x_bytes.len();
+            if y_c1_big > half {
+                x_bytes[num_x_bytes - 1] |= 0x80;
+            } else if y_c1_big == half && y_c0_big > half {
+                x_bytes[num_x_bytes - 1] |= 0x80;
+            }
+            writer.write(&x_bytes)?;
+
+            Ok(())
+        };
+
+        let old_deserialization_logic = |data: &[u8]| -> IoResult<_> {
+            let mut x_bytes_with_y: Vec<u8> = data.to_owned();
+            let x_bytes_with_y_len = x_bytes_with_y.len();
+            let y_over_half = (x_bytes_with_y[x_bytes_with_y_len - 1] & 0x80) == 0x80;
+            x_bytes_with_y[x_bytes_with_y_len - 1] &= 0xFF - 0x80;
+            let x = Fq2::read(x_bytes_with_y.as_slice())?;
+            let x3b = <Bls12_377G2Parameters as SWModelParameters>::add_b(
+                &((x.square() * &x) + &<Bls12_377G2Parameters as SWModelParameters>::mul_by_a(&x)),
+            );
+            let y = x3b.sqrt().ok_or(io::Error::new(
+                io::ErrorKind::NotFound,
+                "couldn't find square root for x",
+            ))?;
+
+            let y_c0_big = y.c0.into_repr();
+            let y_c1_big = y.c1.into_repr();
+
+            let negy = -y;
+
+            let (bigger, smaller) = {
+                let half = Fq::modulus_minus_one_div_two();
+                if y_c1_big > half {
+                    (y, negy)
+                } else if y_c1_big == half && y_c0_big > half {
+                    (y, negy)
+                } else {
+                    (negy, y)
+                }
+            };
+
+            let chosen_y = if y_over_half { bigger } else { smaller };
+            Ok((x, chosen_y))
+        };
```

Comment:
Just to be clear, this logic here will _still_ be buggy for the mentioned edge case where `c1 = 0`?

Reply:
Yes. I'll add a comment